probe-rs/src/probe/wlink/mod.rs

```diff
@@ -0,0 +1,606 @@
+            Err(WchLinkError::CommandFailed(rxbuf[1]).into())
+        }
+    }
+
+    fn dmi_op_write(&mut self, addr: u8, data: u32) -> Result<(u8, u32, u8), DebugProbeError> {
+        let mut rxbuf = [0u8; 9];
+        let raw_data = data.to_be_bytes();
+        self.device.write(
+            &[
+                0x81,
+                0x08,
+                0x06,
+                addr,
+                raw_data[0],
+                raw_data[1],
+                raw_data[2],
+                raw_data[3],
+                DMI_OP_WRITE,
+            ],
+            &mut rxbuf,
+            Duration::from_millis(300),
+        )?;
+
+        let data_out = u32::from_be_bytes([rxbuf[4], rxbuf[5], rxbuf[6], rxbuf[7]]);
+        if rxbuf[0] == 0x82 {
+            Ok((rxbuf[3], data_out, rxbuf[8]))
+        } else {
+            Err(WchLinkError::CommandFailed(rxbuf[1]).into())
+        }
+    }
+
+    fn dmi_op_nop(&mut self, addr: u8, data: u32) -> Result<(u8, u32, u8), DebugProbeError> {
+        let mut rxbuf = [0u8; 9];
+        let raw_data = data.to_be_bytes();
+        self.device.write(
+            &[
+                0x81,
+                0x08,
+                0x06,
+                addr,
+                raw_data[0],
+                raw_data[1],
+                raw_data[2],
+                raw_data[3],
+                DMI_OP_NOP,
+            ],
+            &mut rxbuf,
+            Duration::from_millis(300),
+        )?;
+
+        let data_out = u32::from_be_bytes([rxbuf[4], rxbuf[5], rxbuf[6], rxbuf[7]]);
+        if rxbuf[0] == 0x82 {
+            Ok((rxbuf[3], data_out, rxbuf[8]))
+        } else {
+            todo!()
+        }
+    }
+
+    pub fn assert_unprotected(&mut self) -> Result<(), DebugProbeError> {
+        let rxbuf = self
+            .device
+            .write_command(&[0x81, 0x06, 0x01, 0x01], Duration::from_millis(300))?;
+
+        if rxbuf[0] == 0x82 {
+            if rxbuf[3] == 0x02 {
+                Ok(())
+            } else if rxbuf[3] == 0x01 {
+                Err(WchLinkError::ReadProtected.into())
+            } else {
+                unreachable!()
+            }
+        } else {
+            Err(WchLinkError::CommandFailed(rxbuf[1]).into())
+        }
+    }
+}
+
+impl DebugProbe for WchLink {
+    fn new_from_selector(
+        selector: impl Into<DebugProbeSelector>,
+    ) -> Result<Box<Self>, DebugProbeError>
+    where
+        Self: Sized,
+    {
+        let device = WchLinkUsbDevice::new_from_selector(selector)?;
+        let mut wlink = Self {
+            device,
+            name: "WCH-Link".into(),
+            variant: WchLinkVariant::Ch549,
+            v_major: 0,
+            v_minor: 0,
+            chip_type: 0,
+            riscvchip: 0,
+            last_dmi_read: None,
+        };
+
+        wlink.init()?;
+
+        Ok(Box::new(wlink))
+    }
+
+    fn get_name(&self) -> &str {
+        &self.name
+    }
+
+    fn speed_khz(&self) -> u32 {
+        todo!()
```

Comment:
If this is really not possible for this probe, we will have to adapt the trait. We shouldn't panic here.

Reply:
Yes, it doesn't have a `speed` setting in khz.
Instead, it has a clock setting choice from High, low, and middle. This part of the protocol is not clear for now.